accepted/2020/linking-libraries.md

```diff
@@ -0,0 +1,152 @@
+
+Examples of potential new feature switches:
+- ResourceManager
+    - ResourceManager has logic that can load types from the resource file. This could be conditionally disabled by a feature switch. It is already disabled when loading a `.resource` file directly from disk, which is not embedded in a resource assembly.
+    - Another possibility would be a switch to make exception/resource messages smaller in the libraries. See https://github.com/dotnet/runtime/issues/34057 for more information.
+- EventSource
+    - On some devices the EventSource logic is unnecessary because it has no place to write the events. We could introduce a feature switch to remove EventSource code/calls.
+
+#### Removing Embedded Descriptor Files
+
+There are places in the .NET Libraries where we use private Reflection to call between assemblies. One reason for this is layering. If Assembly A references Assembly B, and Assembly B needs to call a method in Assembly A, it can't reference it directly. We've solved that with a Reflection call.
+
+In order to ensure the linker doesn't remove the private method, we've added "Embedded Descriptor Files" (or `ILLinkTrim.xml` in the `dotnet/runtime` repo) which get embedded into the built assembly. These tell the linker not to trim the specified method/type. The problem is that the linker never trims the specified method/type, even if the callsite of the Reflection is not needed.
+
+See https://github.com/dotnet/runtime/issues/31712 and https://github.com/dotnet/runtime/issues/35199 for more information on this issue.
+
+To solve this, we will follow the approaches outlined in https://github.com/dotnet/runtime/issues/31712 to not rely on the Embedded Descriptor Files in order to make the Reflection usages linker safe. Instead, we will add the appropriate linker annotations and not embed these descriptor files into our libraries. This will allow applications to safely call, or not call, the APIs and the linker will either preserve or trim the code correctly.
+
+There are 2 existing cases here worth mentioning:
+1. Types only used for Debugging
+2. Types used for COM support
+
+In both of these cases, the code is sometimes needed, and sometimes isn't. If you are never going to debug your application (for example, the application is deployed to a device that can't be debugged), it isn't necessary to keep these types in the application. A similar situation exists with COM. Since COM is not IL, the linker isn't able to see if some code in the application is going to use COM or not.
+
+To solve these, we can define a [`feature switch`](#Adding-appropriate-feature-switches) for each of these scenarios. An application (or SDK) can default these to on or off, as appropriate. For example, a Xamarin iOS application can safely turn COM support off.
```

Comment:
Are the COM-related APIs kept around today even on Unix because of the S.P.Corelib ILLinkTrim.xml file?  If yes, will that matter for the goaled-application types anyway, since they'll be targeting the mono runtime and its Corelib?

Separately, if we are thinking about coreclr as part of this effort, have we given any thought to how we might untangle the dependencies introduced by the native runtime pieces?  We generate an ILLinkTrim.xml file that roots anything that might ever be used from native code, regardless of whether that code could actually end up being used.

Reply:
There are COM-only types outside of CoreLib:

https://github.com/dotnet/runtime/blob/e77572ffccc566186f47207f3c5475533c87538e/src/libraries/Microsoft.CSharp/src/ILLinkTrim.xml#L8

https://github.com/dotnet/runtime/blob/e77572ffccc566186f47207f3c5475533c87538e/src/libraries/System.Runtime.InteropServices/src/ILLinkTrim.xml#L3-L4

> have we given any thought to how we might untangle the dependencies introduced by the native runtime pieces?

My only thoughts have been the feature switch in the ILLinkTrim.xml file idea above that would allow application to remove them dynamically. Maybe @vitek-karas or @marek-safar have more thoughts here?

Reply:
> any thought to how we might untangle the dependencies introduced by the native runtime pieces?

Life more of the unmanaged logic into C# as necessary, use explicitly passed in callbacks instead of calling functions by name from VM.

Reply:
> have we given any thought to how we might untangle the dependencies introduced by the native runtime pieces?

It depends. The XML descriptors are actually a good fit for this. If we know the runtime code will be 99% time present we just add it to XML (e.g. string constructor) for the rest it's a combination of using `DynamicDependencyAttribute` (aka PreserveDependency) and for large chunks of native code moving them to manage for better likability.

Reply:
> The XML descriptors are actually a good fit for this

I'm surprised to hear you say that. From what I've seen, the XML isn't a great fit because a non-trivial amount of what's automatically kept isn't actually used by any given app.

Reply:
> by the native runtime pieces

I was specifically referring to core runtime dependencies. There is a certain amount of core types/methods which runtime always requires (e.g. System.Object, System.Exception, etc) to even start. I don't think we need to make that conditional and listing such core dependencies in S.P.C XML descriptor is fine.

Reply:
Sure, there's a subset that's always used.